# AVIATION EXPANSION PROJECT

## Description

Mawingu Airlines has recently enjoyed a run of success, and now we feel that now is the right time to expand our ventures, specifically in the aviation industry. This project seeks to establish whether that path is viable at this point in time by conducting a thorough analysis of data related to various aircraft models. We will determine which aircraft are the lowest risk for the company to start this new business endeavor and then translate your findings into actionable insights that the head of the new aviation division can use to help decide which aircraft to purchase.

## Methodology

The data in focus is contained in the *Aviation dataset* from the National Transportation Safety Board that includes aviation accident data from 1962 to 2023 about civil aviation accidents and selected incidents in the United States and international waters. Our key areas of focus will be three areas:
- **Data Cleaning and Imputation**: We will perform key cleaning techniques on our data and filling in missing values
-  **Data Analysis**: Analysis of key metrics needed for insights into the business
-  **Data Vizualizations**: Graphical vizualization of the key metrics

## 1. Data Cleaning and Imputation

As explained earlier, our data is contained in the *Aviation Dataset* which we will load using the *Pandas* library. This will make viewing our data easier and enable cleaning of the data to be fast and efficient. As always, we will import pandas using the standard alias and read it into our notebook.

In [1]:
import pandas as pd
Aviation_data = pd.read_csv('Aviation_Data.csv')
Aviation_data.head(20)

C:\Users\USER\AppData\Local\Temp\ipykernel_8048\1316027294.py:2: DtypeWarning: Columns (6,7,28) have mixed types. Specify dtype option on import or set low_memory=False.
  Aviation_data = pd.read_csv('Aviation_Data.csv')


Event.Id Investigation.Type Accident.Number  Event.Date  \
0   20001218X45444           Accident      SEA87LA080  1948-10-24   
1   20001218X45447           Accident      LAX94LA336  1962-07-19   
2   20061025X01555           Accident      NYC07LA005  1974-08-30   
3   20001218X45448           Accident      LAX96LA321  1977-06-19   
4   20041105X01764           Accident      CHI79FA064  1979-08-02   
5   20170710X52551           Accident      NYC79AA106  1979-09-17   
6   20001218X45446           Accident      CHI81LA106  1981-08-01   
7   20020909X01562           Accident      SEA82DA022  1982-01-01   
8   20020909X01561           Accident      NYC82DA015  1982-01-01   
9   20020909X01560           Accident      MIA82DA029  1982-01-01   
10  20020909X01559           Accident      FTW82DA034  1982-01-01   
11  20020909X01558           Accident      ATL82DKJ10  1982-01-01   
12  20020917X02148           Accident      FTW82FRJ07  1982-01-02   
13  20020917X02134           Accident      FTW82FRA14  1982-01-02   
14  20020917X02119           Accident      FTW82FPJ10  1982-01-02   
15  20020917X02117           Accident      FTW82FPG08  1982-01-02   
16  20020917X01962           Accident      DEN82DTM08  1982-01-02   
17  20020917X01656           Accident      ANC82FAG14  1982-01-02   
18  20020917X02481           Accident      NYC82DA016  1982-01-02   
19  20020917X02339           Accident      MIA82DA028  1982-01-02   

            Location        Country   Latitude  Longitude Airport.Code  \
0    MOOSE CREEK, ID  United States        NaN        NaN          NaN   
1     BRIDGEPORT, CA  United States        NaN        NaN          NaN   
2      Saltville, VA  United States  36.922223 -81.878056          NaN   
3         EUREKA, CA  United States        NaN        NaN          NaN   
4         Canton, OH  United States        NaN        NaN          NaN   
5         BOSTON, MA  United States  42.445277 -70.758333          NaN   
6         COTTON, MN  United States        NaN        NaN          NaN   
7        PULLMAN, WA  United States        NaN        NaN          NaN   
8   EAST HANOVER, NJ  United States        NaN        NaN          N58   
9   JACKSONVILLE, FL  United States        NaN        NaN          JAX   
10         HOBBS, NM  United States        NaN        NaN          NaN   
11      TUSKEGEE, AL  United States        NaN        NaN          NaN   
12         HOMER, LA  United States        NaN        NaN          NaN   
13        HEARNE, TX  United States        NaN        NaN          T72   
14     CHICKASHA, OK  United States        NaN        NaN          NaN   
15   LITTLE ROCK, AR  United States        NaN        NaN          NaN   
16        MIDWAY, UT  United States        NaN        NaN          NaN   
17       SKWENTA, AK  United States        NaN        NaN          NaN   
18       GALETON, PA  United States        NaN        NaN          5G6   
19         MIAMI, FL  United States        NaN        NaN          NaN   

          Airport.Name  ... Purpose.of.flight Air.carrier  \
0                  NaN  ...          Personal         NaN   
1                  NaN  ...          Personal         NaN   
2                  NaN  ...          Personal         NaN   
3                  NaN  ...          Personal         NaN   
4                  NaN  ...          Personal         NaN   
5                  NaN  ...               NaN  Air Canada   
6                  NaN  ...          Personal         NaN   
7   BLACKBURN AG STRIP  ...          Personal         NaN   
8              HANOVER  ...          Business         NaN   
9    JACKSONVILLE INTL  ...          Personal         NaN   
10                 NaN  ...          Personal         NaN   
11            TUSKEGEE  ...          Personal         NaN   
12                 NaN  ...          Personal         NaN   
13    HEARNE MUNICIPAL  ...          Personal         NaN   
14                 NaN  ...          Personal         NaN   
15                 NaN  ...  

As you can see, pandas has indicated that some columns have mixed data types therefore making it difficult for pandas to inerpret them. We can set the parameter ***low_memory=False*** in our ***pd.read_csv*** function.

In [2]:
import pandas as pd
Aviation_data = pd.read_csv('Aviation_Data.csv', low_memory=False)
Aviation_data.head(20)

Event.Id Investigation.Type Accident.Number  Event.Date  \
0   20001218X45444           Accident      SEA87LA080  1948-10-24   
1   20001218X45447           Accident      LAX94LA336  1962-07-19   
2   20061025X01555           Accident      NYC07LA005  1974-08-30   
3   20001218X45448           Accident      LAX96LA321  1977-06-19   
4   20041105X01764           Accident      CHI79FA064  1979-08-02   
5   20170710X52551           Accident      NYC79AA106  1979-09-17   
6   20001218X45446           Accident      CHI81LA106  1981-08-01   
7   20020909X01562           Accident      SEA82DA022  1982-01-01   
8   20020909X01561           Accident      NYC82DA015  1982-01-01   
9   20020909X01560           Accident      MIA82DA029  1982-01-01   
10  20020909X01559           Accident      FTW82DA034  1982-01-01   
11  20020909X01558           Accident      ATL82DKJ10  1982-01-01   
12  20020917X02148           Accident      FTW82FRJ07  1982-01-02   
13  20020917X02134           Accident      FTW82FRA14  1982-01-02   
14  20020917X02119           Accident      FTW82FPJ10  1982-01-02   
15  20020917X02117           Accident      FTW82FPG08  1982-01-02   
16  20020917X01962           Accident      DEN82DTM08  1982-01-02   
17  20020917X01656           Accident      ANC82FAG14  1982-01-02   
18  20020917X02481           Accident      NYC82DA016  1982-01-02   
19  20020917X02339           Accident      MIA82DA028  1982-01-02   

            Location        Country   Latitude   Longitude Airport.Code  \
0    MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1     BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2      Saltville, VA  United States  36.922223  -81.878056          NaN   
3         EUREKA, CA  United States        NaN         NaN          NaN   
4         Canton, OH  United States        NaN         NaN          NaN   
5         BOSTON, MA  United States  42.445277  -70.758333          NaN   
6         COTTON, MN  United States        NaN         NaN          NaN   
7        PULLMAN, WA  United States        NaN         NaN          NaN   
8   EAST HANOVER, NJ  United States        NaN         NaN          N58   
9   JACKSONVILLE, FL  United States        NaN         NaN          JAX   
10         HOBBS, NM  United States        NaN         NaN          NaN   
11      TUSKEGEE, AL  United States        NaN         NaN          NaN   
12         HOMER, LA  United States        NaN         NaN          NaN   
13        HEARNE, TX  United States        NaN         NaN          T72   
14     CHICKASHA, OK  United States        NaN         NaN          NaN   
15   LITTLE ROCK, AR  United States        NaN         NaN          NaN   
16        MIDWAY, UT  United States        NaN         NaN          NaN   
17       SKWENTA, AK  United States        NaN         NaN          NaN   
18       GALETON, PA  United States        NaN         NaN          5G6   
19         MIAMI, FL  United States        NaN         NaN          NaN   

          Airport.Name  ... Purpose.of.flight Air.carrier  \
0                  NaN  ...          Personal         NaN   
1                  NaN  ...          Personal         NaN   
2                  NaN  ...          Personal         NaN   
3                  NaN  ...          Personal         NaN   
4                  NaN  ...          Personal         NaN   
5                  NaN  ...               NaN  Air Canada   
6                  NaN  ...          Personal         NaN   
7   BLACKBURN AG STRIP  ...          Personal         NaN   
8              HANOVER  ...          Business         NaN   
9    JACKSONVILLE INTL  ...          Personal         NaN   
10                 NaN  ...          Personal         NaN   
11            TUSKEGEE  ...          Personal         NaN   
12                 NaN  ...          Personal         NaN   
13    HEARNE MUNICIPAL  ...          Personal         NaN   
14                 NaN  ...          Personal         NaN   
15      

Great! We have managed to get a small overview of our data. One can see that some columns contain missing values. We will have to see the extent of our missing valies in the data.

In [3]:
Aviation_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90348 entries, 0 to 90347
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      90348 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50132 non-null  object 
 9   Airport.Name            52704 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87507 non-null  object 
 14  Make                    88826 non-null

In [4]:
# Get total number of missing values per column
missing_values = Aviation_data.isnull().sum()

# Get percentage of missing values per column
missing_percentage = (Aviation_data.isnull().sum() / len(Aviation_data)) * 100

# Create a summary DataFrame
missing_data = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_percentage
})

# Display columns with missing values (filter out zeros)
missing_data[missing_data['Missing Values'] > 0].sort_values('Percentage (%)', ascending=False)

Missing Values  Percentage (%)
Schedule                         77766       86.073848
Air.carrier                      73700       81.573471
FAR.Description                  58325       64.555939
Aircraft.Category                58061       64.263736
Longitude                        55975       61.954886
Latitude                         55966       61.944924
Airport.Code                     40216       44.512330
Airport.Name                     37644       41.665560
Broad.phase.of.flight            28624       31.681941
Publication.Date                 16689       18.471909
Total.Serious.Injuries           13969       15.461327
Total.Minor.Injuries             13392       14.822686
Total.Fatal.Injuries             12860       14.233851
Engine.Type                       8555        9.468942
Report.Status                     7843        8.680878
Purpose.of.flight                 7651        8.468367
Number.of.Engines                 7543        8.348829
Total.Uninjured                   7371        8.158454
Weather.Condition                 5951        6.586753
Aircraft.damage                   4653        5.150086
Registration.Number               2841        3.144508
Injury.Severity                   2459        2.721698
Country                           1685        1.865011
Amateur.Built                     1561        1.727764
Model                             1551        1.716695
Make                              1522        1.684597
Location                          1511        1.672422
Accident.Number                   1459        1.614867
Event.Date                        1459        1.614867
Event.Id                          1459        1.614867

Wow! The table above us shows that there are a significant number of columns with missing values. This will be problematic for us since it will hamper with our analysis and our vizualiations later in the project. Let us now categorize our columns in terms of missing values.

### Critical Missing (>50% missing):
- Schedule (86.07%)

- Air.carrier (81.57%)

- FAR.Description (64.56%)

- Aircraft.Category (64.26%)

- Longitude (61.95%)

- Latitude (61.94%)

### Moderate Missing (10-50% missing):
- Airport.Code (44.51%)

- Airport.Name (41.67%)

- Broad.phase.of.flight (31.68%)

### Low Missing (<10% missing):
- All other columns

For our data, we can see that we have six columns that contain a significant percentage of missing values within the columns. For the three topmost columns, since they are missing more than 80% of data and likely won't provide meaningful analysis even with imputation, we can consider dropping them.

### 1. Critical missing columns

In [5]:
columns_to_drop = ['Schedule', 'Air.carrier', 'FAR.Description']
Aviation_data = Aviation_data.drop(columns=columns_to_drop)

For the columns with location data and moderate percentage of missing data, we can conduct data imputation and fill them with the relevant summary statistics, i.e., mean, median, e.t.c. , or we can decide to drop them since we have the country columns.

### 2(a). Aircraft Category:

In [6]:
Aviation_data['Aircraft.Category'] = Aviation_data['Aircraft.Category'].fillna('Unknown')

### 2(b). Airport.code, Airport.name and Broad.phase.of.flight:

We can perform mode imputation on these columns since dropping them will prove harmful to our analysis:

In [7]:
Aviation_data['Airport.Code'] = Aviation_data['Airport.Code'].fillna(Aviation_data['Airport.Code'].mode()[0])
Aviation_data['Airport.Name'] = Aviation_data['Airport.Name'].fillna(Aviation_data['Airport.Name'].mode()[0])
Aviation_data['Broad.phase.of.flight'] = Aviation_data['Broad.phase.of.flight'].fillna(Aviation_data['Broad.phase.of.flight'].mode()[0])

### 2(c). Latitude and longitude columns:

In [8]:
Aviation_data['Latitude'] = pd.to_numeric(Aviation_data['Latitude'], errors='coerce')  # 'coerce' turns invalid values to NaN
Aviation_data['Longitude'] = pd.to_numeric(Aviation_data['Longitude'], errors='coerce')

Since we can see that the location columns contain a high percentage of missing values, we can just decide to drop them since they will hamper us with our analysis:

In [9]:
Aviation_data.drop(columns=['Latitude', 'Longitude'], inplace=True)

### 3. Low missing columns:

Lets have a look at how our dataset looks now.

In [10]:
Aviation_data.head(20)

Event.Id Investigation.Type Accident.Number  Event.Date  \
0   20001218X45444           Accident      SEA87LA080  1948-10-24   
1   20001218X45447           Accident      LAX94LA336  1962-07-19   
2   20061025X01555           Accident      NYC07LA005  1974-08-30   
3   20001218X45448           Accident      LAX96LA321  1977-06-19   
4   20041105X01764           Accident      CHI79FA064  1979-08-02   
5   20170710X52551           Accident      NYC79AA106  1979-09-17   
6   20001218X45446           Accident      CHI81LA106  1981-08-01   
7   20020909X01562           Accident      SEA82DA022  1982-01-01   
8   20020909X01561           Accident      NYC82DA015  1982-01-01   
9   20020909X01560           Accident      MIA82DA029  1982-01-01   
10  20020909X01559           Accident      FTW82DA034  1982-01-01   
11  20020909X01558           Accident      ATL82DKJ10  1982-01-01   
12  20020917X02148           Accident      FTW82FRJ07  1982-01-02   
13  20020917X02134           Accident      FTW82FRA14  1982-01-02   
14  20020917X02119           Accident      FTW82FPJ10  1982-01-02   
15  20020917X02117           Accident      FTW82FPG08  1982-01-02   
16  20020917X01962           Accident      DEN82DTM08  1982-01-02   
17  20020917X01656           Accident      ANC82FAG14  1982-01-02   
18  20020917X02481           Accident      NYC82DA016  1982-01-02   
19  20020917X02339           Accident      MIA82DA028  1982-01-02   

            Location        Country Airport.Code        Airport.Name  \
0    MOOSE CREEK, ID  United States         NONE             Private   
1     BRIDGEPORT, CA  United States         NONE             Private   
2      Saltville, VA  United States         NONE             Private   
3         EUREKA, CA  United States         NONE             Private   
4         Canton, OH  United States         NONE             Private   
5         BOSTON, MA  United States         NONE             Private   
6         COTTON, MN  United States         NONE             Private   
7        PULLMAN, WA  United States         NONE  BLACKBURN AG STRIP   
8   EAST HANOVER, NJ  United States          N58             HANOVER   
9   JACKSONVILLE, FL  United States          JAX   JACKSONVILLE INTL   
10         HOBBS, NM  United States         NONE             Private   
11      TUSKEGEE, AL  United States         NONE            TUSKEGEE   
12         HOMER, LA  United States         NONE             Private   
13        HEARNE, TX  United States          T72    HEARNE MUNICIPAL   
14     CHICKASHA, OK  United States         NONE             Private   
15   LITTLE ROCK, AR  United States         NONE             Private   
16        MIDWAY, UT  United States         NONE         FIELD RANCH   
17       SKWENTA, AK  United States         NONE             Private   
18       GALETON, PA  United States          5G6      CHERRY SPRINGS   
19         MIAMI, FL  United States         NONE             Private   

   Injury.Severity Aircraft.damage  ...    Engine.Type Purpose.of.flight  \
0         Fatal(2)       Destroyed  ...  Reciprocating          Personal   
1         Fatal(4)       Destroyed  ...  Reciprocating          Personal   
2         Fatal(3)       Destroyed  ...  Reciprocating          Personal   
3         Fatal(2)       Destroyed  ...  Reciprocating          Personal   
4         Fatal(1)       Destroyed  ...            NaN          Personal   
5        Non-Fatal     Substantial  ...      Turbo Fan               NaN   
6         Fatal(4)       Destroyed  ...  Reciprocating          Personal   
7        Non-Fatal     Substantial  ...  Reciprocating          Personal   
8        Non-Fatal     Substantial  ...  Reciprocating          Business   
9        Non-Fatal     Substantial  ...  Reciprocating          Personal   
10       Non-Fatal     Substantial  ...  Reciprocating          Personal   
11       Non-Fatal     Substantial  ...  Reciprocating          Personal   
12       Non-Fatal       Destroyed  ...  Reciprocating     

We have managed to drop 5 columns. Lets go ahead and work on the remaining columns.

### 3(a). Injury columns

For the injury columns, you notice that the severity column indicates the extent of the injuries recorded in a particular crash. One can also notice thatsome rows have for example *fatal(2)* and it corresponds to the total fatal injuries column. However some have *Fatal* but you find that both total fatal and total serious columns having inputs leaving one to question whether the severity and the various injury columns influence each other. We will need to perform some cleaning and create a new column to sort the mix-up

In [11]:
# Check unique values in Injury.Severity to define rules
print(Aviation_data['Injury.Severity'].unique())

['Fatal(2)' 'Fatal(4)' 'Fatal(3)' 'Fatal(1)' 'Non-Fatal' 'Incident'
 'Fatal(8)' 'Fatal(78)' 'Fatal(7)' 'Fatal(6)' 'Fatal(5)' 'Fatal(153)'
 'Fatal(12)' 'Fatal(14)' 'Fatal(23)' 'Fatal(10)' 'Fatal(11)' 'Fatal(9)'
 'Fatal(17)' 'Fatal(13)' 'Fatal(29)' 'Fatal(70)' 'Unavailable'
 'Fatal(135)' 'Fatal(31)' 'Fatal(256)' 'Fatal(25)' 'Fatal(82)'
 'Fatal(156)' 'Fatal(28)' 'Fatal(18)' 'Fatal(43)' 'Fatal(15)' 'Fatal(270)'
 'Fatal(144)' 'Fatal(174)' 'Fatal(111)' 'Fatal(131)' 'Fatal(20)'
 'Fatal(73)' 'Fatal(27)' 'Fatal(34)' 'Fatal(87)' 'Fatal(30)' 'Fatal(16)'
 'Fatal(47)' 'Fatal(56)' 'Fatal(37)' 'Fatal(132)' 'Fatal(68)' 'Fatal(54)'
 'Fatal(52)' 'Fatal(65)' 'Fatal(72)' 'Fatal(160)' 'Fatal(189)'
 'Fatal(123)' 'Fatal(33)' 'Fatal(110)' 'Fatal(230)' 'Fatal(97)'
 'Fatal(349)' 'Fatal(125)' 'Fatal(35)' 'Fatal(228)' 'Fatal(75)'
 'Fatal(104)' 'Fatal(229)' 'Fatal(80)' 'Fatal(217)' 'Fatal(169)'
 'Fatal(88)' 'Fatal(19)' 'Fatal(60)' 'Fatal(113)' 'Fatal(143)' 'Fatal(83)'
 'Fatal(24)' 'Fatal(44)' 'Fatal(64)' 'Fatal(92

In [12]:
# Rule 1: If "Fatal(1)", set Total.Fatal.Injuries=1, others=0
fatal_mask = Aviation_data['Injury.Severity'].str.contains('Fatal', na=False)
Aviation_data.loc[fatal_mask, 'Total.Fatal.Injuries'] = 1.0
Aviation_data.loc[fatal_mask, ['Total.Serious.Injuries', 'Total.Minor.Injuries']] = 0.0

# Rule 2: If "Non-Fatal", set Total.Minor.Injuries=1, others=0
non_fatal_mask = Aviation_data['Injury.Severity'].str.contains('Non-Fatal', na=False)
Aviation_data.loc[non_fatal_mask, 'Total.Minor.Injuries'] = 1.0
Aviation_data.loc[non_fatal_mask, ['Total.Fatal.Injuries', 'Total.Serious.Injuries']] = 0.0

In [13]:
missing_severity = Aviation_data['Injury.Severity'].isna() #checking for missing values
has_minor_injuries = (Aviation_data['Total.Minor.Injuries'] > 0)
Aviation_data.loc[missing_severity & has_minor_injuries, 'Injury.Severity'] = 'Non-Fatal(1)'

In [14]:
# Check Fatal cases
print(Aviation_data[fatal_mask][['Injury.Severity', 'Total.Fatal.Injuries', 'Total.Serious.Injuries']].head())

# Check Non-Fatal cases
print(Aviation_data[non_fatal_mask][['Injury.Severity', 'Total.Minor.Injuries', 'Total.Serious.Injuries']].head())

  Injury.Severity  Total.Fatal.Injuries  Total.Serious.Injuries
0        Fatal(2)                   1.0                     0.0
1        Fatal(4)                   1.0                     0.0
2        Fatal(3)                   1.0                     0.0
3        Fatal(2)                   1.0                     0.0
4        Fatal(1)                   1.0                     0.0
   Injury.Severity  Total.Minor.Injuries  Total.Serious.Injuries
5        Non-Fatal                   1.0                     0.0
7        Non-Fatal                   1.0                     0.0
8        Non-Fatal                   1.0                     0.0
9        Non-Fatal                   1.0                     0.0
10       Non-Fatal                   1.0                     0.0


We have now resolved the conflict between the severity and the various injury count columns, but we still haven't dealt with the missing values within the columns. Let's do so.

In [18]:
# Rule: If "Fatal(1)", fill NaN in Total.Fatal.Injuries with 1, others with 0
Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Fatal', na=False), 'Total.Fatal.Injuries'] = (
    Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Fatal', na=False), 'Total.Fatal.Injuries'].fillna(1)
)
Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Fatal', na=False), ['Total.Serious.Injuries', 'Total.Minor.Injuries']] = (
    Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Fatal', na=False), ['Total.Serious.Injuries', 'Total.Minor.Injuries']].fillna(0)
)

# Rule: If "Non-Fatal(1)", fill NaN in Total.Minor.Injuries with 1, others with 0
Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Non-Fatal', na=False), 'Total.Minor.Injuries'] = (
    Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Non-Fatal', na=False), 'Total.Minor.Injuries'].fillna(1)
)
Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Non-Fatal', na=False), ['Total.Fatal.Injuries', 'Total.Serious.Injuries']] = (
    Aviation_data.loc[Aviation_data['Injury.Severity'].str.contains('Non-Fatal', na=False), ['Total.Fatal.Injuries', 'Total.Serious.Injuries']].fillna(0)
)

# For rows with missing Injury.Severity, fill all injury NaN with 0 (or median if preferred)
injury_cols = ['Total.Fatal.Injuries', 'Total.Serious.Injuries', 'Total.Minor.Injuries']
no_severity_mask = Aviation_data['Injury.Severity'].isna()
Aviation_data.loc[no_severity_mask, injury_cols] = Aviation_data.loc[no_severity_mask, injury_cols].fillna(0)

In [20]:
# Fill NaN with 0 (assuming no uninjured people if data is missing)
Aviation_data['Total.Uninjured'] = Aviation_data['Total.Uninjured'].fillna(0).astype(int)

### 3(b). Event date and Publication date:

For these columns, we have to first convert them to date times and then fill in the missing values

In [23]:
# Convert to datetime (handles most common string formats automatically)
Aviation_data['Event.Date'] = pd.to_datetime(Aviation_data['Event.Date'], errors='coerce', dayfirst=True)
Aviation_data['Publication.Date'] = pd.to_datetime(Aviation_data['Publication.Date'], errors='coerce', dayfirst=True)
print(f"Event.Date missing after conversion: {Aviation_data['Event.Date'].isna().mean():.1%}")
print(f"Publication.Date missing after conversion: {Aviation_data['Publication.Date'].isna().mean():.1%}")

Event.Date missing after conversion: 1.6%
Publication.Date missing after conversion: 18.5%


Let us now handle them individually

#### For Event.Date (1.7% missing):

In [25]:
# Fill with a logical estimate (e.g., publication date minus median delay)
if not Aviation_data['Publication.Date'].isna().all():
    median_delay = (Aviation_data['Publication.Date'] - Aviation_data['Event.Date']).median()
    Aviation_data['Event.Date'] = Aviation_data['Event.Date'].fillna(Aviation_data['Publication.Date'] - median_delay)

#### For Publication.Date (19% missing):

In [26]:
# Option 1: Fill with event date plus median publication delay (if Event.Date exists)
median_delay = (Aviation_data['Publication.Date'] - Aviation_data['Event.Date']).median()
Aviation_data['Publication.Date'] = Aviation_data['Publication.Date'].fillna(Aviation_data['Event.Date'] + median_delay)

In [28]:
# Check if any publication dates predate event dates
invalid = Aviation_data[Aviation_data['Publication.Date'] < Aviation_data['Event.Date']]
print(f"{len(invalid)} illogical records (published before event)")

# Optionally correct these cases
Aviation_data.loc[Aviation_data['Publication.Date'] < Aviation_data['Event.Date'], 'Publication.Date'] = Aviation_data['Event.Date'] + pd.Timedelta(days=7)

2 illogical records (published before event)


Great! we have dealt with two more columns! We can now deal with the remaining columns.

#### 4. Remaining Columns

#### 4(a). Categorical Columns (Low Missingness < 10%):

In [33]:
categorical_cols = [
    'Engine.Type', 'Report.Status', 'Purpose.of.flight', 
    'Weather.Condition', 'Aircraft.damage', 'Country',
     'Model', 'Make', 'Location'
]

# Fill with mode (most frequent category) per column
for col in categorical_cols:
    Aviation_data[col] = Aviation_data[col].fillna(Aviation_data[col].mode()[0])

# For binary categories (e.g., 'Amateur.Built'), consider:
Aviation_data['Amateur.Built'] = Aviation_data['Amateur.Built'].fillna('Unknown')

#### 4(b). Numeric Columns:

In [35]:
# Fill with median (robust to outliers)
Aviation_data['Number.of.Engines'] = Aviation_data['Number.of.Engines'].fillna(Aviation_data['Number.of.Engines'].median())

#### 4(c). Identifier Columns:

In [36]:
# Option 1: Fill with "UNKNOWN" (if IDs are critical for tracking)
Aviation_data['Registration.Number'] = Aviation_data['Registration.Number'].fillna('UNKNOWN_REG')
Aviation_data['Accident.Number'] = Aviation_data['Accident.Number'].fillna('UNKNOWN_ACC')
Aviation_data['Event.Id'] = Aviation_data['Event.Id'].fillna('UNKNOWN_EVT')

Now we have dealt with the missing values effectively. We can now get to the second phase of our project.

## Data Analysis